source/initial_temperature/patch_on_S40RTS.cc

```diff
@@ -0,0 +1,197 @@
+        // function
+        Assert (false, ExcNotImplemented());
+
+      // scale the perturbation in seismic velocity into a density perturbation
+      // vs_to_density is an input parameter
+      const double density_perturbation = vs_to_density * vs_perturbation;
+
+      double temperature_perturbation;
+      if (depth > s40rts.no_perturbation_depth)
+        // scale the density perturbation into a temperature perturbation
+        temperature_perturbation =  -1./s40rts.thermal_alpha * density_perturbation;
+      else
+        // set heterogeneity to zero down to a specified depth
+        temperature_perturbation = 0.0;
+
+      return background_temperature + temperature_perturbation;
+    }
+
+    template <int dim>
+    void
+    PatchOnS40RTS<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection ("Initial temperature model");
+      {
+        Utilities::AsciiDataBase<dim>::declare_parameters(prm,
+                                                          "$ASPECT_SOURCE_DIR/data/initial-temperature/patch-on-S40RTS/test/",
+                                                          "upper_shell_3d.txt");
+        prm.declare_entry ("Maximum grid depth", "700000.0",
+                           Patterns::Double (0),
+                           "The maximum depth of the Vs ascii grid. The model will read in  "
+                           "Vs from S40RTS below this depth.");
+        prm.declare_entry ("Smoothing length scale", "200000.0",
+                           Patterns::Double (0),
+                           "The depth range (above maximum grid depth) over which to smooth. "
+                           "Smoothing is done by averaging the values of the ascii grid and S40RTS at each point.");
+
+      }
+      prm.leave_subsection();
+    }
+
+    template <int dim>
+    void
+    PatchOnS40RTS<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Initial temperature model");
+      {
+        Utilities::AsciiDataBase<dim>::parse_parameters(prm);
+
+        max_grid_depth           = prm.get_double ("Maximum grid depth");
+        smoothing_length_scale   = prm.get_double ("Smoothing length scale");
+
+      }
+      prm.leave_subsection ();
+      s40rts.initialize_simulator (this->get_simulator());
+
+      // Note: parse_parameters will call initialize for us
+      s40rts.parse_parameters(prm);
+    }
+  }
+}
+
+// explicit instantiations
+namespace aspect
+{
+  namespace InitialTemperature
+  {
+    ASPECT_REGISTER_INITIAL_TEMPERATURE_MODEL(PatchOnS40RTS,
+                                              "patch on S40RTS",
+                                              "Implementation of a model in which the initial "
+                                              "temperature is derived from a file containing data "
+                                              "in ascii format (e.g. a high resolution upper mantle tomography) "
```

Comment:
Please add a line about the type of data in the file, e.g. seismic velocity perturbations.